server/src/main/java/com/distkv/server/metaserver/client/DmetaClient.java

```diff
@@ -1,87 +1,61 @@
+  public DmetaClient(String confStr) throws TimeoutException, InterruptedException {
 
-  public DmetaClient() {
+    final Configuration conf = JRaftUtils.getConfiguration(confStr);
 
-    final Configuration conf = new Configuration();
-    if (!conf.parse(confStr)) {
-      throw new IllegalArgumentException("Fail to parse conf:" + confStr);
-    }
-
-    RouteTable.getInstance().updateConfiguration(groupId, conf);
+    RouteTable.getInstance().updateConfiguration(GROUP_ID, conf);
 
     //init RPC client and update Routing table
     cliClientService = new BoltCliClientService();
     cliClientService.init(new CliOptions());
-  }
-
-  public void put(String key, String value) {
-    try {
-      if (!RouteTable.getInstance().refreshLeader(cliClientService, groupId, 1000).isOk()) {
-        throw new IllegalStateException("Refresh leader failed");
-      }
-      //get leader term
-      final PeerId leader = RouteTable.getInstance().selectLeader(groupId);
-
-      PutRequest request = new PutRequest();
-      request.setKey(key);
-      request.setValue(value);
-      PutResponse response = (PutResponse) cliClientService.getRpcClient()
-          .invokeSync(leader.getEndpoint().toString(), request, 3000);
-      if (!response.isSuccess()) {
-        throw new RuntimeException("put error");
-      }
-    } catch (InterruptedException e) {
-      e.printStackTrace();
-    } catch (TimeoutException e) {
-      e.printStackTrace();
-    } catch (RemotingException e) {
-      e.printStackTrace();
+    //refresh leader term.
+    if (!RouteTable.getInstance().refreshLeader(cliClientService, GROUP_ID, 1000).isOk()) {
+      throw new IllegalStateException("Refresh leader failed");
     }
   }
 
-  public String get(String key) {
+  public HeartbeatResponse heartbeat(NodeInfo nodeInfo, int timeout) {
     try {
-      if (!RouteTable.getInstance().refreshLeader(cliClientService, groupId, 1000).isOk()) {
-        throw new IllegalStateException("Refresh leader failed");
-      }
-      //get leader term
-      final PeerId leader = RouteTable.getInstance().selectLeader(groupId);
 
-      final GetRequest request = new GetRequest();
-      request.setKey(key);
+      //get leader term.
+      final PeerId leader = RouteTable.getInstance().selectLeader(GROUP_ID);
 
-      GetResponse response = (GetResponse) cliClientService.getRpcClient()
-          .invokeSync(leader.getEndpoint().toString(), request, 3000);
+      final HeartbeatRequest request = new HeartbeatRequest(nodeInfo);
+
+      HeartbeatResponse response = (HeartbeatResponse) cliClientService.getRpcClient()
+          .invokeSync(leader.getEndpoint().toString(), request, timeout);
       if (!response.isSuccess()) {
-        throw new RuntimeException("get error");
+        if (response.getRedirect().length() > 0) {
+          PeerId peerId = new PeerId();
+          peerId.parse(response.getRedirect());
+          RouteTable.getInstance().updateLeader(GROUP_ID, peerId);
```

Comment:
These cases should be reset the leaser:
1. heartbeat timeout.
2. redirect leader.